src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/DataFactoryCustomActivity.cs

```diff
@@ -120,43 +125,43 @@ private async Task AnonymizeBlobInJsonFormatAsync(BlobClient inputBlobClient, Bl
+            FhirStreamReader reader = new FhirStreamReader(inputStream);
+            FhirBlobConsumer consumer = new FhirBlobConsumer(outputBlobClient);
+            Func<string, string> anonymizerFunction = (item) =>
+            {
+                try
                 {
-                    try
-                    {
-                        resultLine = _engine.AnonymizeJson(resourceLine);
-                        writer.WriteLine(resultLine);
-                        processedCount += 1;
-                    }
-                    catch (Exception innerException)
-                    {
-                        processedErrorCount += 1;
-                        Console.WriteLine($"[{blobName}]: Anonymize partial failed, you can find detail error message in stderr.txt.");
-                        Console.Error.WriteLine($"[{blobName}]: Error #{processedErrorCount}\nResource: {resourceLine}\nErrorMessage: {innerException.Message}\n Details: {innerException.ToString()}\nStackTrace: {innerException.StackTrace}");
-                    }
+                    return _engine.AnonymizeJson(item);
                 }
-                writer.Flush();
-            }
+                catch (Exception ex)
+                {
+                    Console.WriteLine($"[{blobName}]: Anonymize partial failed, you can find detail error message in stderr.txt.");
+                    Console.Error.WriteLine($"[{blobName}]: Resource: {item}\nErrorMessage: {ex.Message}\n Details: {ex.ToString()}\nStackTrace: {ex.StackTrace}");
+                    throw;
+                }
+            };
 
-            using (FileStream uploadFileStream = new FileStream(outputTempFileName, FileMode.Open, FileAccess.Read, FileShare.None, 4096, FileOptions.DeleteOnClose))
+            Stopwatch stopWatch = Stopwatch.StartNew();
+            FhirPartitionedExecutor executor = new FhirPartitionedExecutor(reader, consumer, anonymizerFunction);
+            executor.PartitionCount = Environment.ProcessorCount * 2;
+            Progress<BatchAnonymizeProgressDetail> progress = new Progress<BatchAnonymizeProgressDetail>();
+            progress.ProgressChanged += (obj, args) =>
             {
-                await outputBlobClient.UploadAsync(uploadFileStream);
-                Console.WriteLine($"[{blobName}]: Succeeded in {processedCount} resources, failed in {processedErrorCount} resources in total.");
-            }
+                Interlocked.Add(ref processedCount, args.ProcessCompleted);
+                Interlocked.Add(ref processedErrorCount, args.ProcessFailed);
+                Interlocked.Add(ref consumedCount, args.ConsumeCompleted);
+
+                Console.WriteLine($"[{stopWatch.Elapsed.ToString()}][tid:{args.CurrentThreadId}]: {processedCount} Completed. {processedErrorCount} Failed. {consumedCount} consume completed.");
+            };
+
+            await executor.ExecuteAsync(CancellationToken.None, false, progress).ConfigureAwait(false);
```

Comment:
Can/Should we build resiliency against node crash in the middle of a large blob processing (say picking only unprocessed data on restart)?

Reply:
Maybe we can add file level retry at ADF pipeline first? Use foreach activity to run anonymizer tool on single file and retry if it failed?

To achieve resume functionality, looks like we need somewhere for partial status, in batch mode, we can do this with additional storage table or ... If we change to use azure function, looks like we can easily leverage the partial status.

Reply:
Should we add a story to track this work?

Reply:
Create a story to backlog for tracking: https://microsofthealth.visualstudio.com/Health/_workitems/edit/73121/